### ISSUE TITLE
[討論]非線性回歸模型與最小平方法

### ISSUE THREAD

**Issue by kuihao** [andy6804tw/2021-13th-ironman#1]
請問非線性模型是否可以用最小平方法來解?
若是可以，為何沒有人用最小平方法來解非線性的logistic regression?

**Comment by andy6804tw** [comment]
感謝 @kangweilin  補充！
邏輯回歸可以稱迴歸的線性分類器，因此 logistic regression 我們會說他是一個線性分類器。因為迴歸分析中 Y 是已經觀察到的資料，可是邏輯迴歸中 P(Y=1|X) 是資料裡面無法觀察到的，所以我們就沒辦法用傳統的最小平方法估計，而是要採用最大概似法 (Maximum Likelihood Estimation)。
關於更多數學公式以及上面講的的專有名詞可以參考台大李宏毅[ML Lecture 5: Logistic Regression](https://www.youtube.com/watch?v=hSXFuypLukA)。

**Comment by kuihao** [comment]
謝謝 @kangweilin  和 @andy6804tw  的回覆，我後來再次回顧李宏毅教授[16年](https://youtu.be/fZAZUYEeIMg?t=777)及 [去年的影片 (類神經網路訓練不起來怎麼辦 (四)：損失函數 (Loss) 也可能有影響)](https://youtu.be/O2VkP8dJ5FE?t=661)後，算是有想通這個問題的解答。

我原本想問的是高中所教的 Close Form 版本的最小平方法解法 (將 Loss Function 轉換成矩陣求解)，可以直接套用在 Linear Regression，而大學數值分析方法也教到線性回歸可以把參數換成非線性函數，並同樣可用反矩陣解。既然如此，作為同樣是非線性函數的 Logistic Regression 是否同樣存在一個 Close Form 解法呢？

然而，我忽略的是 Logistic Regression 的非線性函數 (Activation Functon) 是在Features與Weights進行 Weighted Sum 之後才套用 (  f(x_i, w_i) = sigmoid(w_1*x_1 + w_2*x_2 + ... + w_n*x_n)+bias )，而非單純對每個 Features 套用 Activation Functon 才加總( f(x_i, w_i) = sigmoid(w_1)*x_1 + sigmoid(w_2)*x_2 + ... + sigmoid(w_n)*x_n)+bias )，這樣的差異導致計算偏導函數的時候無法像傳統的回歸方程式整理出 Close Form。

有趣的是，把 Logistic Regression 的 Loss Function 換成 MSE (平均平方誤差) 仍可使用梯度下降法來進行計算，這部分在 @andy6804tw 推薦的 [ML Lecture 5: Logistic Regression](https://www.youtube.com/watch?v=hSXFuypLukA) 及 [類神經網路訓練不起來怎麼辦 (四)：損失函數 (Loss) 也可能有影響)](https://youtu.be/O2VkP8dJ5FE?t=661) 中，教授各自從不同的觀點來解釋為何 Logistic Regression 搭配 MSE 並不是一個合適的選擇 (即使經過測試，用 Tensorflow 的確可以 Training，而且也不是 Train 不起來，只是不好 Train。)

另外，@andy6804tw 您所提到: 
> 迴歸分析中 Y 是已經觀察到的資料，可是邏輯迴歸中 P(Y=1|X) 是資料裡面無法觀察到的，所以我們就沒辦法用傳統的最小平方法估計。

由於邏輯迴歸也是屬於監督式學習，因此訓練上仍需準備 Label，這應該算是一種觀察到的數據? 若以 MSE 替換 Cross Entropy 硬 Train 一發，就是將 Label 視為一個新的 Feature 維度，作為回歸模型想要擬合的目標，假設是二元分類問題就可以把 Label 製作成 {0, 1}，剛好也會符合 Logistic Regression 最後通過 sigmoid 輸出，使 Model prediction 剛好落在 [0,1]，但如同教授所言，這個方式一來要 Train 很久，二來是造成模型過度在意 Outlier，會使決策邊界訂在不太合理的地方。

最後補個之前在 Youtube 上看到有人用[函數可視化的方式介紹如何從 Maximum Likelihood 推導成 Minimum Cross-Entropy](https://www.youtube.com/watch?v=up85PnWFBY8)，我覺得相當生動，一併附上讓未來其他看到這篇的後輩學習。
